pool/src/lib.rs

```diff
@@ -530,25 +525,76 @@ impl LendingPool {
     fn do_deposit(
         env: &Env,
         who: &Address,
-        s_token_address: &Address,
         asset: &Address,
         amount: i128,
-        collat_accrued_rate: i128,
+        reserve: &ReserveData,
     ) -> Result<bool, Error> {
-        let token = token::Client::new(env, asset);
-        token.transfer(who, s_token_address, &amount);
+        if amount == 0 {
+            return Ok(false);
+        }
 
-        let s_token = s_token_interface::STokenClient::new(env, s_token_address);
+        let collat_coeff = Self::get_collateral_coeff(env, reserve)?;
+        let underlying_asset = token::Client::new(env, asset);
+        let s_token = STokenClient::new(env, &reserve.s_token_address);
         let is_first_deposit = s_token.balance(who) == 0;
 
-        // amount_to_mint = amount / collat_accrued_rate
-        let amount_to_mint = FixedI128::from_inner(collat_accrued_rate)
+        let amount_to_mint = collat_coeff
             .recip_mul_int(amount)
             .ok_or(Error::MathOverflowError)?;
+
+        underlying_asset.transfer(who, &reserve.s_token_address, &amount);
         s_token.mint(who, &amount_to_mint);
+
+        event::deposit(env, who.clone(), asset.clone(), amount);
+
         Ok(is_first_deposit)
     }
 
+    fn do_repay(
+        env: &Env,
+        who: &Address,
+        asset: &Address,
+        amount: i128,
+        reserve: &ReserveData,
+    ) -> Result<(i128, bool), Error> {
+        let debt_token = DebtTokenClient::new(env, &reserve.debt_token_address);
+        let asset_debt = debt_token.balance(who);
+
+        if asset_debt == 0 {
+            return Ok((amount, false));
+        }
+
+        let underlying_asset = token::Client::new(env, asset);
+        let debt_coeff = Self::get_debt_coeff(env, reserve)?;
+
+        let compounded_debt = debt_coeff
+            .mul_int(asset_debt)
+            .ok_or(Error::MathOverflowError)?;
+
+        let payback_amount = amount.min(compounded_debt);
+
+        let payback_debt = debt_coeff
+            .recip_mul_int(payback_amount)
+            .ok_or(Error::MathOverflowError)?;
+
+        underlying_asset.transfer(who, &reserve.s_token_address, &payback_amount);
+        debt_token.burn(who, &payback_debt);
+
+        event::repay(env, who.clone(), asset.clone(), amount);
+
+        let remaning_amount = amount
+            .checked_sub(asset_debt)
+            .ok_or(Error::MathOverflowError)?;
+
+        let remaning_amount = if remaning_amount > 0 {
+            remaning_amount
+        } else {
+            0
+        };
+
+        Ok((remaning_amount, compounded_debt == payback_amount))
```

Comment:
It is not clear what does expression mean `compounded_debt == payback_amount`. It would be more clear to have extra bool variable like `let is_repayed = compounded_debt == payback_amount;`